modules/parachains/src/lib.rs

```diff
@@ -335,10 +352,83 @@ pub mod pallet {
 			at_relay_block: (RelayBlockNumber, RelayBlockHash),
 			parachains: Vec<(ParaId, ParaHash)>,
 			parachain_heads_proof: ParaHeadsProof,
+		) -> DispatchResultWithPostInfo {
+			Self::submit_parachain_heads_ex(
+				origin,
+				at_relay_block,
+				parachains,
+				parachain_heads_proof,
+				false,
+			)
+		}
+
+		/// Change `PalletOwner`.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::call_index(1)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_owner(origin: OriginFor<T>, new_owner: Option<T::AccountId>) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_owner(origin, new_owner)
+		}
+
+		/// Halt or resume all pallet operations.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::call_index(2)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			operating_mode: BasicOperatingMode,
+		) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_operating_mode(origin, operating_mode)
+		}
+
+		/// Submit proof of one or several parachain heads.
+		///
+		/// The proof is supposed to be proof of some `Heads` entries from the
+		/// `polkadot-runtime-parachains::paras` pallet instance, deployed at the bridged chain.
+		/// The proof is supposed to be crafted at the `relay_header_hash` that must already be
+		/// imported by corresponding GRANDPA pallet at this chain.
+		///
+		/// The call fails if:
+		///
+		/// - the pallet is halted;
+		///
+		/// - the relay chain block `at_relay_block` is not imported by the associated bridge
+		///   GRANDPA pallet.
+		///
+		/// The call may succeed, but some heads may not be updated e.g. because pallet knows
+		/// better head or it isn't tracked by the pallet.
+		///
+		/// The `is_free_execution_expected` parameter is not really used inside the call. It is
```

Comment:
The idea is that by making this a call parameter, we leave an option to submit non-free transactions. Imagine e.g. if relay/parachain headers suddenly grow in size and they're now not accepted for free. We can solve it two ways:
- update `MAX_HEADER_SIZE` constants in a fellowship repo, wait for next release and upgrade bridge hub runtime. For all this time, the bridge will be halted - because signed extension will just reject all paid transactions;
- OTOH we may leave ourselves a backup option of submitting non-free headers. Even though, it would mean losing relayer tokens, we will be able to (at least) submit mandatory GRANPDA headers (which are required for the bridge and its security). We could even continue delivering messages and headers if we consider bridge operations more important than the tokens we spend on fees.

Reply:
Makes sense, would even be nice to have ⬆️ written in some comment too.

Reply:
Everything looks ok so far and yes, I think I understand how that works to allow free and non-free, basically this is the important [IF](https://github.com/paritytech/parity-bridges-common/blob/sv-refund-non-mandatory-headers/modules/parachains/src/call_ext.rs#L72-L75) where `is_free_execution_expected` is really used.

The only thing I have, I can't shake the feeling that this is a bit hacky - to pass `is_free_execution_expected` as a call parameter even it is used just by transaction extension.

What I am trying to say, I think the other way how to pass `is_free_execution_expected` to that [IF](https://github.com/paritytech/parity-bridges-common/blob/sv-refund-non-mandatory-headers/modules/parachains/src/call_ext.rs#L72-L75) is to use a transaction extension parameter instead of a call parameter.

E.g. here we submit finality: https://github.com/paritytech/parity-bridges-common/blob/sv-refund-non-mandatory-headers/relays/lib-substrate-relay/src/finality/target.rs#L106-L130, so instead of modifying `call` we can control this by transaction extension parameter `free_execution_expected`:

```
		let call = P::SubmitFinalityProofCallBuilder::build_submit_finality_proof_call(
			header, proof, context,
		);
		self.client
			.submit_and_watch_signed_extrinsic(
				&self.transaction_params.signer,
				move |best_block_id, transaction_nonce| {
					Ok(UnsignedTransaction::new(call.into(), transaction_nonce)
						.era(TransactionEra::new(best_block_id, mortality)))
						// TODO: we can controll `is_free_execution_expected` here
						.free_execution_expected(Some(true/false))                               
				},
			)
			.await
```

Basically, something like when you pass `era` or `tip` to the signed extensions.
We support custom extensions here: https://github.com/paritytech/parity-bridges-common/blob/sv-refund-non-mandatory-headers/primitives/chain-bridge-hub-cumulus/src/lib.rs#L169-L173

Reply:
> What I am trying to say, I think the other way how to pass is_free_execution_expected to that [IF](https://github.com/paritytech/parity-bridges-common/blob/sv-refund-non-mandatory-headers/modules/parachains/src/call_ext.rs#L72-L75) is to use a transaction extension parameter instead of a call parameter.

That would be a huge pain for all community I suppose - we'll need to talk to all wallets, all tx creation libraries, ... and tell them how to properly craft bridge hub transactions :) Yet you will need to pass this parameter to all other transactions. Even though it is a single byte and maybe size doesn't matter much here, you will need to explain that fact to everyone - e.g. when someone call `balances::transfer()`, he'll probably wants to know what this single byte is about and why he may pass anything

Reply:
That said, I agree that it isn't perfect - pallets code is already splitted between actual calls and extension. In bridges v2 I have been initially planning to remove extensions at all, but looks like they're here forever. But maybe because of that we will be able to remove some parts of the pallet code and rely on existing tx extensions (now some light checks are made twice).

Reply:
Ok, cool, thank you, Slava, for your explanation. In the flash of this explanation, I see this "call parameter passing to transaction extension" as a nice hack (workaround, feature..) :smile:  . So now I can let my 'hacky feeling' go away" and finish review :)